sys/vm/vm_fault.c

```diff
@@ -1751,37 +1751,44 @@ vm_fault_prefault(const struct faultstate *fs, vm_offset_t addra,
  * pages are successfully held, then the number of held pages is returned
  * together with pointers to those pages in the array "ma".  However, if any
  * of the pages cannot be held, -1 is returned.
+ *
+ * XXX-CHERI: In a pure-capability world with physical capabilities we'd
+ * likely want to bound the first and last pages when rounding is required.
  */
 int
-vm_fault_quick_hold_pages(vm_map_t map, vm_offset_t addr, vm_size_t len,
+vm_fault_quick_hold_pages(vm_map_t map, void * __capability addr, vm_size_t len,
     vm_prot_t prot, vm_page_t *ma, int max_count)
 {
-	vm_offset_t end, va;
+	vm_offset_t end, start, va;
 	vm_page_t *mp;
 	int count;
 	boolean_t pmap_failed;
 
 	if (len == 0)
 		return (0);
-	end = round_page(addr + len);
-	addr = trunc_page(addr);
+#if __has_feature(capabilities)
+	if (!__CAP_CHECK(addr, len))
+		return (-1);
+#endif
+	end = (__cheri_addr vm_offset_t)round_page((char * __capability)addr + len);
+	start = (__cheri_addr vm_offset_t)trunc_page(addr);
```

Comment:
```suggestion
	start = (__cheri_addr vm_offset_t)trunc_page(addr);
	end = (__cheri_addr vm_offset_t)round_page((char * __capability)addr + len);
```
Now independent so can be the right way round.